### PR TITLE
Fix DP dispatch padding gradient bias in FSDP training

### PR DESCRIPTION
**Pull request by 0russwest0**
## Summary

Fix gradient bias caused by padded samples introduced for verl DP dispatch.

verl requires `DataProto` batches to be evenly divisible by the target DP dispatch size before splitting them across DP shards. To satisfy this constraint, the trainer pads post-rollout batches before actor / critic updates. These padded samples are dispatch artifacts and should not contribute to training.

Previously, padded samples could still affect loss normalization, mini-batch statistics, or critic value-loss masking in some FSDP / new-engine paths. When the number of real samples was not divisible by the DP dispatch size, different DP shards could observe different effective sample / token counts, causing biased actor or critic gradients.

This PR makes padded samples explicit via `sample_mask`, propagates global mini-batch sample / token statistics, and uses pad-aware Agent-R1-local loss aggregation so losses are normalized by real global training data rather than dispatch padding.

## Changes

- Introduce tensor-based `sample_mask` to explicitly mark samples added only for verl DP dispatch padding.
- Zero out `response_mask` for padded samples so dispatch padding cannot contribute to actor or critic loss.
- Update valid-data filtering to use `sample_mask` instead of the previous non-tensor `is_pad` marker.
- Compute the padding divisor from the actual training dispatch DP sizes instead of treating padding as a generic worker-world-size requirement.
- Add global mini-batch metadata:
  - `mini_batch_id`
  - `mini_batch_global_size`
  - `mini_batch_global_token_num`
- Use these global statistics in pad-aware loss aggregation so loss denominators are based on real samples / tokens, not padded local shards.
- Add zero-denominator handling for empty or fully padded micro-batches.
- Patch actor / critic update paths to skip empty response micro-batches.
- Ensure critic value masks exclude padded samples during critic updates.
- Add Agent-R1-local policy / value loss implementations and worker wrappers to apply the corrected aggregation consistently across legacy FSDP and new-engine paths.
- Patch new-engine static micro-batch splitting to handle uneven local batches while keeping DP ranks synchronized on micro-batch count.

## Notes
- This PR keeps verl's current DP dispatch / `DataProto` chunking behavior and makes the training path pad-aware. As a follow-up, we can consider improving verl's dispatch logic to support uneven DP partitions directly, which should remove the need to pad `DataProto` batches purely for dispatch.
- This PR keeps the value loss aggregation aligned with upstream verl. A follow-up PR will separately audit whether global normalization kwargs should be propagated beyond the policy loss path in the new engine, including `compute_value_loss` and related metric-only aggregations where applicable.
- A separate verl new-engine `compute_values` collection issue was observed for nested value outputs. It appears unrelated to this padding-gradient fix and can be addressed later with a dedicated verl patch.

### CODE REVIEW

File: agent_r1/trainer/ppo/core_algos.py [hunk 1]
```diff
@@ -315,3 +316,254 @@ def compute_grpo_outcome_advantage(
...
+    def is_zero_denom(denom):
+        if torch.is_tensor(denom):
+            return denom.detach().item() == 0
+        return denom == 0
+
+    if loss_agg_mode == "token-mean":
+        if batch_num_tokens is None:
+            denom = loss_mask.sum()
+        else:
+            denom = batch_num_tokens
```
[the remark this answers is not in the document]
Comment (0russwest0) [reply agent_r1/trainer/ppo/core_algos.py:341]: I think this mixes up two fields: `global_token_num` is the list passed by `train_mini_batch`, while `batch_num_tokens` used by `agg_loss` is still a scalar computed from `loss_mask.sum()`. So this should not hit a TypeError on the current path.

File: agent_r1/workers/engine_workers.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+                    global_token_num = global_token_num[global_token_num > 0].tolist()
+                    global_batch_size = int(mini_batch_global_sizes[indices[0]].item())
+
+                    tu.assign_non_tensor(
+                        mini_batch_td,
+                        global_token_num=NonTensorData(global_token_num),
+                        global_batch_size=global_batch_size,
+                        update_lr_scheduler=iteration_idx == total_num_iterations - 1,
+                        disable_auto_offload=True,
+                    )
```
[the remark this answers is not in the document]
Comment (0russwest0) [reply agent_r1/workers/engine_workers.py:155]: I think this is already handled by the engine. `mini_batch_td` does not carry `dp_size` from `TrainingWorker.train_mini_batch`, but before the loss function is called, `forward_backward_batch` injects both `batch_num_tokens` and `dp_size` into `data`. So the loss functions already see `data["dp_size"]` on this path.

File: agent_r1/trainer/ppo/core_algos.py [hunk 3]
```diff
@@ -315,3 +316,254 @@ def compute_grpo_outcome_advantage(
...
+
+
+def compute_value_loss(
+    vpreds: torch.Tensor,
+    returns: torch.Tensor,
+    values: torch.Tensor,
+    response_mask: torch.Tensor,
+    cliprange_value: float,
+    loss_agg_mode: str = "token-mean",
+):
```
[the remark this answers is not in the document]
Comment (0russwest0) [reply agent_r1/trainer/ppo/core_algos.py:386]: Good point. I’ll keep this PR aligned with upstream verl for now, since value loss aggregation currently follows the same behavior there. I’ll note this as follow-up work and audit propagation of global normalization kwargs beyond the new-engine policy loss path in a separate PR.

**Review by lqzxt — APPROVED** [review]
